packages/frontend/src/stdlib/analyses/schema_erd.tsx

```diff
@@ -0,0 +1,280 @@
+    const header = () => (
+        <>
+            <IconButton
+                onClick={() => {
+                    const svg = svgString();
+                    if (svg) {
+                        download(svg, `${schemaName()} - ERD.svg`, "image/svg+xml");
+                    }
+                }}
+                disabled={!svgString()}
+                tooltip={`Export the entity-relationship diagram as SVG`}
+            >
+                <Download size={16} />
+            </IconButton>
+        </>
+    );
+
+    return (
+        <div class="graph-visualization-container">
+            <Foldable title="Entity-relationship diagram" header={header()}>
+                <FormGroup compact>
+                    <SelectField
+                        label="Direction"
+                        value={props.content.direction ?? Direction.Vertical}
+                        onChange={(evt) => {
+                            props.changeContent((content) => {
+                                content.direction = evt.currentTarget.value as Direction;
+                            });
+                        }}
+                    >
+                        <option value={Direction.Horizontal}>{"Horizontal"}</option>
+                        <option value={Direction.Vertical}>{"Vertical"}</option>
+                    </SelectField>
+                </FormGroup>
+            </Foldable>
+            <div class="graph-visualization">
+                <Show when={svgString()}>{(svg) => <div innerHTML={svg()} />}</Show>
+            </div>
+        </div>
+    );
+}
+
+/** Convert a schema model into an ERD-style Graphviz graph using HTML-like labels. */
+export function schemaToERD(model: DblModel): Viz.Graph {
+    const entityType: ObType = { tag: "Basic", content: "Entity" };
+    const attrType: MorType = { tag: "Basic", content: "Attr" };
+
+    const entities = model.obGeneratorsWithType(entityType);
+    const nodes: Required<Viz.Graph>["nodes"] = [];
+
+    // Collect all mappings to know which entities they point to
+    const mappingsByEntity = new Map<
+        string,
+        Array<{ id: string; name: string; targetEntity: string }>
+    >();
+    const mappingType: MorType = { tag: "Hom", content: entityType };
+    for (const morId of model.morGeneratorsWithType(mappingType)) {
+        const mor = model.morPresentation(morId);
+        invariant(mor, `Could not find mapping morphism presentation for id: ${morId}`);
+        invariant(
+            mor.dom.tag === "Basic",
+            `Unexpected morphism domain. Expecting 'Basic' got: ${mor.dom.tag}`,
+        );
+        invariant(
+            mor.cod.tag === "Basic",
+            `Unexpected morphism co-domain. Expecting 'Basic' got: ${mor.dom.tag}`,
+        );
+        invariant(entities.includes(mor.dom.content), "Domain content not in entities");
+        invariant(entities.includes(mor.cod.content), "Codomain content not in entities");
```

Comment:
Since this code is using the elaborated model, but regardless of whether it validates, the model *can* be ill-typed if the user has made a mistake. Thus, the safe thing is to replace these invariants with conditionals of the form

```typescript
if (!(...)) {
    continue;
}
```

Alternatively, you could only show the visualization when the model validates, in which case it would be correct to use invariants, but IMO it's better to make visualization analyses degrade gracefully in the face of user error when possible.

Reply:
Thanks. Yeah, I actually thought this code would trigger errors and was surprised when it didn't during my testing. I think this has to do with recent changes to the elaborator for models of discrete theories (now using DoubleTT rather than the previous, more direct method) which is causing invalid arrows to be filtered out during elaboration when they weren't before. However, this behavior should be considered accidental and isn't guaranteed to persist. The general rule is that you can't assume a model to be valid/well-typed unless it passes validation.